### PR TITLE
Fix MLX save and GGUF export parity issues

### PR DESCRIPTION
**Pull request by Lyxot**
This PR fixes several issues around MLX merged saves and GGUF export parity.

Bug list:

1. VLM merged/full saves were missing VLM config fidelity
   Root cause: MLX merged saves used the text-only `mlx_lm.utils.save_config` path for VLM checkpoints, so VLM-specific config structure, normalization, and fields were not preserved in `config.json`.
   Commit: 615aff8e

2. QLoRA merged 16-bit saves were not fully dequantized
   Root cause: The merged 16-bit path treated LoRA fusion with `fuse(dequantize=True)` as sufficient dequantization, but quantized non-LoRA modules could remain packed.
   Commit: 55fa1737

3. VLM GGUF mmproj output was metadata-only / missing real projector tensors
   Root cause: The temporary GGUF staging directory contained mlx-vlm sanitized tensor names/layouts, while llama.cpp MMPROJ conversion expects HF/llama.cpp projector names/layouts; in-place rewriting also risked corrupting file-backed tensors.
   Commit: 6fcaa19a

4. MLX GGUF export options like `first_conversion` were dropped by the bound save method
   Root cause: The monkey-patched MLX `save_pretrained_gguf` wrapper accepted CUDA-compatible `**kwargs` but discarded them when calling the utility implementation, so caller-specified GGUF options like `first_conversion` never reached export.
   Commit: 0029c545

5. VLM GGUF exports could leave same-name vision tensors in MLX layout, e.g. Gemma3 patch embeddings stayed OHWI instead of OIHW
   Root cause: VLM sanitizer inversion only considered renamed candidates; for models whose sanitizer leaves a tensor name unchanged but changes layout, replay could accept the unchanged name/tensor and stop before applying the needed layout transform.
   Commit: 97c980fc

6. VLM GGUF sanitizer replay failed for models whose sanitizers need real submodules, e.g. GLM-OCR
   Root cause: Sanitizer replay used a fake proxy object with only `config`/`args`, but model sanitizers such as GLM-OCR call submodule sanitizers through real attributes like `self.vision_tower`, so replay failed and no tensor rewrite happened.
   Commit: c7002e78

7. VLM GGUF export could fail with modern package-layout llama.cpp converters due missing `conversion/` sibling package
   Root cause: Modern llama.cpp conversion scripts are package entrypoints that import sibling `conversion.*` modules; writing/running the patched script from a detached cache directory broke that package layout, and the MLX path could also use a different llama.cpp tree than the shared patcher root.
   Commit: 4546225d

8. MLX VLM loading could degrade to tokenizer-only processors, so GGUF staging omitted image processor metadata needed by mmproj conversion
   Root cause: GLM-OCR custom processor construction could fail through Transformers `AutoImageProcessor` without `torchvision`, and mlx-vlm fell back to tokenizer-only metadata.
   Commit: 398cb9ca

9. VLM GGUF staging could omit image processor sidecars like `preprocessor_config.json`, breaking mmproj conversion
   Root cause: MLX merged saves trusted tokenizer/model saves to emit every non-weight sidecar needed by downstream converters, but degraded or tokenizer-only processors can omit source files such as `preprocessor_config.json`, `processor_config.json`, `video_preprocessor_config.json`, tokenizer model files, or custom processing files.
   Commit: df5deb90

10. GLM-OCR GGUF exports could advertise a NextN/MTP layer that the MLX model did not export, making llama.cpp load fail on missing `blk.16.*` tensors
    Root cause: Source config metadata could advertise speculative/NextN/MTP layers through fields such as `num_nextn_predict_layers`, `nextn_predict_layers`, or `mtp_num_hidden_layers`, but the loaded/exported MLX language model did not contain those extra blocks.
    Commit: d6d5d39b

11. Raw mlx-vlm model saves could miss `config.json` because `model.config` dataclass configs were not extracted
    Root cause: `_get_model_config` only handled `_config` dictionaries and `model.args`, while raw mlx-vlm models can expose config as a dataclass-like `model.config`.
    Commit: 962dec85

Validation:
- Ran `python -m pytest tests/test_mlx_save_export_regressions.py -q`.
- Result: 23 passed.

### CODE REVIEW

**Comment by danielhanchen** [comment]
Triple-confirmed this PR end-to-end:

## 1. Local Linux, four Python versions

47 tests pass on each of 3.10 / 3.11 / 3.12 / 3.13: the 23 regressions from `tests/test_mlx_save_export_regressions.py` plus 24 additional edge-case probes I wrote to cover backward-compat negatives and the Copilot review comments (`_copy_source_sidecars` with non-directory src, `_has_vision_config` with malformed `thinker_config`, `_read_json_file` with binary garbage and `PermissionError`, `_rewrite_mlx_vlm_tensor_for_gguf` empty pipeline, `_save_mlx_config` text-only routing, `_is_vlm_model` on primitives, etc.). All green on every interpreter.

## 2. Cross-OS shim path

`tests/test_mlx_save_export_regressions.py` + `tests/test_mlx_torch_shim_smoke.py` green on macos-14 (Apple Silicon), ubuntu-latest, and windows-latest CI runners. The shim path catches Linux / Windows pathlib regressions that real Apple Silicon hardware would not surface.

## 3. Real Apple Silicon mlx wheels plus Studio boot

Drove `bash install.sh --local --no-torch`, force-reinstalled this PR's `unsloth-zoo` head into the Studio venv, installed real `mlx 0.x` / `mlx-lm` / `mlx-vlm 0.5.0` wheels, and ran 18 contract probes against the live Apple Silicon kernels (one per fix). All 18 PASS. Then booted `unsloth studio` (API-only) with the PR's `unsloth-zoo` overlaid: `/api/health` returned `{\"status\":\"healthy\", ...}` in 23 seconds. No Studio boot regression.

## Pre-PR vs post-PR diff

Copied this PR's test file onto `unslothai/unsloth-zoo:main` and re-ran it: **22 failed, 1 passed**. Same suite on the PR head: **23 passed**. The single pre-PR passer is `test_lora_push_uses_lora_adapter_hub_path`, a defensive guard the author added to confirm the unchanged LoRA-only push path stays intact -- it correctly passes on both sides. Every other test detects a real, observable failure on main. Sample failure on main: `TypeError: push_to_hub_gguf() got an unexpected keyword argument 'first_conversion'`, which is exactly the bug fix #4 addresses.

## Coverage per fix

| # | Commit | Confirmed |
|---|---|---|
| 1 | `615aff8e` | VLM config save uses `mlx_vlm.utils.save_config`; `quantization_config` preserved. Text-only still routes through `mlx_lm.utils.save_config` (backward-compat negative) |
| 2 | `55fa1737` | QLoRA merged 16-bit fully dequantizes non-LoRA quantized modules |
| 3 | `6fcaa19a` | VLM GGUF mmproj tensor names/layouts normalized |
| 4 | `0029c545` | `save_pretrained_gguf` / `push_to_hub_gguf` accept and forward `first_conversion` (was the smoking gun on main) |
| 5 | `97c980fc` | Same-name layout transforms applied; `_mlx_arrays_match` value-checks real rank-2 mlx arrays |
| 6 | `c7002e78` | `_MlxVlmSanitizeProxy` constructable; replay uses real submodules |
| 7 | `4546225d` | Patched converter co-located with `conversion/` sibling |
| 8 | `398cb9ca` | `_repair_degraded_vlm_processor` rebuilds from sidecars; `_read_json_file` returns `{}` for missing / binary; propagates unexpected errors |
| 9 | `df5deb90` | `_copy_source_sidecars` copies non-weight sidecars, skips weights, handles non-directory src cleanly |
| 10 | `d6d5d39b` | NextN / MTP metadata stripped when MLX language model doesn't export those layers |
| 11 | `962dec85` | `_get_model_config` extracts dataclass-shaped `model.config` |

## Notes for the maintainer

- The branch is currently based on `606d2b2d`, which is one commit behind main's `bea8b483 tool mask support (#688)`. A two-way `git diff main..HEAD` renders `unsloth_zoo/rl_replacements.py` as removed; a real three-way merge preserves those additions because this branch never touches the file. Worth a rebase before merge to make the diff easier to read.
- Unrelated to this PR: `tests/test_mlx_finetune_last_n_layers.py::test_get_peft_model_passes_finetune_last_n_layers_through` fails identically on `main` and on this branch with `AttributeError: 'FakeModel' object has no attribute 'trainable_parameters'`. Separate cleanup.
- Also unrelated to this PR: `.github/workflows/mlx-ci.yml` on `main` still imports the pre-subpackage flat paths (`unsloth_zoo.mlx_loader` etc.). It will fail the moment this PR's subpackage layout becomes the default. Should be refreshed to `unsloth_zoo.mlx.{loader,compile,utils,trainer,cce}`.
- Copilot-flagged minor concerns the PR has not yet addressed (none of these block merge): `_mlx_arrays_match` value-blind on ranks 3 and 6 plus; `os.environ` mutation in `_download_convert_hf_to_gguf_cached` is process-global / not thread-safe; `isinstance(value, mx.array)` -- works fine on the shim and on real mlx 0.x, just flagging for future-proofing.

LGTM.

**Comment by danielhanchen** [comment]
Pushed `5a6ee091` onto this branch with one small fix and a larger edge case test suite I wrote while stress testing the PR.

## Bug fix: `_read_json_file` returned non-dict values

A sidecar holding valid JSON that is not an object (`[]`, `"text"`, `null`, `3.14`) was returned as is, and `_repair_degraded_vlm_processor` then crashed on `processor_config.get("processor_class")` with `AttributeError: 'list' object has no attribute 'get'`. So a malformed `processor_config.json` could break VLM loading inside the very function meant to repair degraded processors. The fix coerces non-dict payloads to `{}`, matching the documented contract and the existing handling for missing or unparseable files.

## New tests: `tests/test_mlx_save_export_edge_cases.py` (71 tests)

Named with the `test_mlx_` prefix so the MLX CI glob picks it up. It covers the branches the 23 regression tests do not reach:

- `_mlx_arrays_match` on rank 1/3/6 tensors, scalars, identity shortcut, backend errors, NaN. This also settles the review concern about value-blind matching on rank 3 plus: current code value checks all ranks.
- `_rewrite_mlx_vlm_tensor_for_gguf`: 5D layout transforms, the 1D floating `tensor - 1` candidate, multi-pipeline inputs, raising sanitizers, `model.` prefixed alias families.
- `_MlxVlmSanitizeProxy` via a two-argument `sanitize(self, weights)` class method (previously zero coverage on that path).
- `_prepare_vlm_gguf_export_directory` end to end with real safetensors shards: rewrite plus atomic replace, `index.json` weight_map remapping, duplicate-name RuntimeError, missing index, unicode shard names, and all NextN decrement/skip branches including `language_config` and `thinker_config.text_config`.
- `_copy_source_sidecars`: src equals dst, symlinks, suffixless and uppercase-suffix files, unicode names, spaces in paths, permission errors.
- `_is_vlm_model`, `_get_model_config`, `_save_mlx_config` real bodies (explicit flags, dataclass and `to_dict` fallbacks, quantization key precedence).
- `save_pretrained_gguf`: default `first_conversion` derivation through the bf16 intermediate plus quantize step and intermediate cleanup; six threads exporting concurrently to prove the patcher env lock serializes and restores `UNSLOTH_LLAMA_CPP_SCRIPTS_DIR`, including a preexisting user override.
- Converter placement contract: package layout patched script lands beside `conversion/`, monolith layout stays in `LLAMA_CPP_DEFAULT_DIR` (backward-compat guard for the `llama_cpp.py` change).

One shim adjustment was needed: `mx.save_safetensors` in `tests/mlx_simulation` now normalizes tensors to contiguous, since real MLX has no contiguity constraint but `safetensors.torch` rejects non-contiguous views. Without it the real shard rewrite loop was untestable on the shim.

## Validation

- This branch: `pytest tests/test_mlx_*.py` gives 273 passed, 1 skipped. The single failure is the preexisting `test_mlx_finetune_last_n_layers.py` FakeModel issue, identical on `main`, fix already open in #700.
- Same result across isolated venvs on Python 3.10, 3.11, 3.12 and 3.13 with CPU torch 2.12, and transformers 4.51.3, 4.57.6 and 5.5.0.
- Comment-only polish was verified by AST comparison (docstrings stripped) against the tested tree, so the published code is exactly what ran.

No behavior changes outside the one loader fix. The branch still merges clean onto current `main`.

**Comment by danielhanchen** [comment]
Pushed `50309fe7`, a comment and docstring only pass over this PR's diff: collapsed multi line blocks, removed comments that restate the code, kept the load bearing rationale (the file backed safetensors gotcha, converter placement reasoning, processor repair context). Net 45 lines removed.

Scope is restricted to lines this PR adds, so it cannot conflict with the comment tightening already merged on main. Every touched file was verified code identical to the previous commit by AST comparison with docstrings stripped, and the full MLX suite still passes (273 passed, 1 skipped, plus the known pre existing failure tracked in #700).

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/mlx/utils.py [hunk 1]
```diff
@@ -4203,6 +4720,7 @@ def push_to_hub_gguf(
     save_directory,
     repo_id,
     quantization_method="fast_quantized",
+    first_conversion=None,
     token=None,
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/utils.py]: Valid catch. Fixed in 350d7900: first_conversion now sits after token and private, so the pre-existing positional form push_to_hub_gguf(model, tok, dir, repo, method, token) keeps binding token as token. Guarded by test_push_to_hub_gguf_positional_token_stays_token.

File: unsloth_zoo/mlx/loader.py [hunk 2]
```diff
@@ -360,6 +360,154 @@ def _load_weights_without_projection_quant_state(self, file_or_weights, strict=T
...
+def _resolve_mlx_vlm_processor_class(model_type, processor_class_name):
+    """Resolve a custom mlx-vlm or Transformers processor class by name."""
+    if not processor_class_name:
+        return None
+
+    module_model_type = (model_type or "").replace("-", "_")
+    module_candidates = (
+        f"mlx_vlm.models.{module_model_type}.processing",
+        f"mlx_vlm.models.{module_model_type}.processing_{module_model_type}",
+    )
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/loader.py:396]: Agreed. 350d7900 adds the MODEL_REMAPPING target package to the module candidates, so aliased model types resolve their custom processor classes before falling back to Transformers. Covered by test_resolve_processor_class_follows_model_remapping.

File: unsloth_zoo/mlx/utils.py [hunk 3]
```diff
@@ -3441,6 +3480,465 @@ def _get_src_path(model):
...
+    if module is None or getattr(module, "sanitize", None) is None:
+        return
+    if all(existing is not module for existing, _ in steps):
+        steps.append((module, None))
+
+
+def _get_mlx_vlm_model_sanitize_pipelines(model):
+    """Build sanitizer pipelines from a loaded mlx-vlm model and submodules."""
+    if model is None or getattr(model, "sanitize", None) is None:
+        return []
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/utils.py]: Agreed. 350d7900 makes the wrapper step optional: submodule sanitizers such as vision_tower.sanitize now build replay pipelines even when the wrapper has no sanitize of its own, so those exports get the HF-layout rewrite. Covered by test_sanitize_pipelines_built_from_submodule_only_sanitizers plus a guard that the wrapper-first ordering is unchanged when the wrapper does sanitize.

File: unsloth_zoo/mlx/loader.py [hunk 4]
```diff
@@ -360,6 +360,154 @@ def _load_weights_without_projection_quant_state(self, file_or_weights, strict=T
...
+            import transformers
+            image_processor_class = getattr(transformers, image_processor_type, None)
+            if image_processor_class is not None:
+                return image_processor_class(**image_kwargs)
+        except Exception:
+            pass
+
+    try:
+        from transformers import AutoImageProcessor
+        return AutoImageProcessor.from_pretrained(model_path)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/loader.py:451]: Agreed. 350d7900 tries the mlx-vlm processing module for the named image processor class after the Transformers lookup and before the AutoImageProcessor retry, so mlx-vlm native image processors can be rebuilt from sidecar kwargs. Covered by test_image_processor_rebuild_uses_mlx_vlm_module_class.

**Comment by danielhanchen** [comment]
Ran a post-merge verification on this since it landed mid-review: the parity fixes are artifact-level correct, not just call-level. Confirmed `mlx_lm.utils.save_config` pops `vision_config` (so routing VLMs through `mlx_vlm.utils.save_config` is the right fix, and the `quantization_config` mirroring matches), `dequantize_model` genuinely dequantizes standalone quantized non-LoRA modules and strips quant metadata, and the VLM GGUF shard rewrite keeps `model.safetensors.index.json` consistent, with the write-beside-then-replace pattern protecting mmap-backed tensors. 99/99 of the PR's tests pass on the Linux CPU shim, the sibling MLX suites (57 tests) still pass, and CUDA-path risk is nil (lazy imports; the one shared `llama_cpp.py` change defaults to byte-identical legacy behavior). Apple Silicon CI on main including this change is green: https://github.com/unslothai/unsloth-zoo/actions/runs/27282789929

One small follow-up worth a one-liner: the "already saved" check in `push_to_hub_merged` (`unsloth_zoo/utils.py` around line 4612) does not account for single-file `model.safetensors`, causing a redundant re-save. Non-blocking.
